File: python/cbact01c/cbact01c.py
"""Python port of batch program CBACT01C.

Reproduces the COBOL procedure flow of ``app/cbl/CBACT01C.cbl``:

    * log  "START OF EXECUTION OF PROGRAM CBACT01C"
    * open the account master (here: a DB session)
    * read every account sequentially in ACCT-ID ascending order
    * print each record's labelled fields (paragraph 1100-DISPLAY-ACCT-RECORD)
    * log  "END OF EXECUTION OF PROGRAM CBACT01C"

Error handling mirrors the COBOL file-status logic: normal cursor exhaustion is
end-of-file (a clean exit), while any unexpected DB/read error is reported like
paragraph 9910-DISPLAY-IO-STATUS and the program exits non-zero to mirror the
9999-ABEND-PROGRAM / CEE3ABD abend.

This replaces the JCL job ``app/jcl/READACCT.jcl``.
"""

from __future__ import annotations

import sys
from typing import Callable, Iterable, Iterator

from sqlalchemy import select
from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.orm import Session

from cbact01c.account import Account
from cbact01c.db import create_db_engine, create_session_factory

PROGRAM_ID = "CBACT01C"

# Non-zero return code used to mirror the COBOL abend (9999-ABEND-PROGRAM).
ABEND_RETURN_CODE = 12

# Ordered (label, accessor) pairs, matching paragraph 1100-DISPLAY-ACCT-RECORD
# exactly: same labels, same order. Each label is 25 characters wide (COBOL
# concatenates the literal and the value with no separator).
_MONEY = lambda v: f"{v:.2f}"  # noqa: E731 - render Decimal with 2 dp
_FIELDS: list[tuple[str, Callable[[Account], str]]] = [
    ("ACCT-ID                 :", lambda a: f"{a.acct_id:011d}"),
    ("ACCT-ACTIVE-STATUS      :", lambda a: a.acct_active_status),
    ("ACCT-CURR-BAL           :", lambda a: _MONEY(a.acct_curr_bal)),
    ("ACCT-CREDIT-LIMIT       :", lambda a: _MONEY(a.acct_credit_limit)),
    ("ACCT-CASH-CREDIT-LIMIT  :", lambda a: _MONEY(a.acct_cash_credit_limit)),
    ("ACCT-OPEN-DATE          :", lambda a: a.acct_open_date),
    ("ACCT-EXPIRAION-DATE     :", lambda a: a.acct_expiraion_date),
    ("ACCT-REISSUE-DATE       :", lambda a: a.acct_reissue_date),
    ("ACCT-CURR-CYC-CREDIT    :", lambda a: _MONEY(a.acct_curr_cyc_credit)),
    ("ACCT-CURR-CYC-DEBIT     :", lambda a: _MONEY(a.acct_curr_cyc_debit)),
    ("ACCT-GROUP-ID           :", lambda a: a.acct_group_id),
]

SEPARATOR = "-" * 49


def format_account_record(account: Account) -> list[str]:
    """Return the labelled output lines for one account (paragraph 1100)."""
    lines = [f"{label}{accessor(account)}" for label, accessor in _FIELDS]
    lines.append(SEPARATOR)
    return lines


def display_account_record(account: Account, out: Callable[[str], None]) -> None:
    """Print one account's labelled fields, mirroring 1100-DISPLAY-ACCT-RECORD."""
    for line in format_account_record(account):
        out(line)


def iter_accounts(session: Session) -> Iterator[Account]:
    """Yield every account ordered by ACCT-ID ascending, streaming the cursor.

    Uses ``yield_per`` so records are fetched incrementally, matching the COBOL
    sequential READ loop rather than loading the whole file into memory.
    """
    stmt = select(Account).order_by(Account.acct_id.asc())
    yield from session.scalars(stmt).yield_per(100)


def run_batch(
    accounts: Iterable[Account],
    out: Callable[[str], None] = print,
    err: Callable[[str], None] = lambda m: print(m, file=sys.stderr),
) -> int:
    """Run the CBACT01C procedure over an iterable of accounts.

    Returns 0 on success. On an unexpected read error, reports the failure like
    9910-DISPLAY-IO-STATUS / 9999-ABEND-PROGRAM and returns a non-zero code.
    """
    out(f"START OF EXECUTION OF PROGRAM {PROGRAM_ID}")
    try:
        # Normal exhaustion of the iterator is end-of-file: a clean exit.
        for account in accounts:
            display_account_record(account, out)
    except SQLAlchemyError as exc:
        err("ERROR READING ACCOUNT FILE")
        _display_io_status(exc, err)
        _abend(err)
        return ABEND_RETURN_CODE
    out(f"END OF EXECUTION OF PROGRAM {PROGRAM_ID}")
    return 0


def _display_io_status(exc: Exception, err: Callable[[str], None]) -> None:
    """Report an I/O failure, mirroring paragraph 9910-DISPLAY-IO-STATUS."""
    err(f"FILE STATUS IS: {exc}")


def _abend(err: Callable[[str], None]) -> None:
    """Emit the abend banner, mirroring paragraph 9999-ABEND-PROGRAM."""
    err("ABENDING PROGRAM")


def main(argv: list[str] | None = None) -> int:
    """CLI entry point. Opens a DB session and runs the batch."""
    engine = create_db_engine()
    session_factory = create_session_factory(engine)
    try:
        with session_factory() as session:
            return run_batch(iter_accounts(session))
    except SQLAlchemyError as exc:
        # Failure opening the "file" (establishing the session/connection).
        print("ERROR OPENING ACCTFILE", file=sys.stderr)
        _display_io_status(exc, lambda m: print(m, file=sys.stderr))
        _abend(lambda m: print(m, file=sys.stderr))
        return ABEND_RETURN_CODE


if __name__ == "__main__":
    raise SystemExit(main())

File: modern/src/main/java/com/cog/carddemo/Cbact01c.java
package com.cog.carddemo;

import java.io.IOException;
import java.io.PrintStream;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * Modern equivalent of the batch program {@code app/cbl/CBACT01C.cbl}: read the
 * account master file sequentially and print every account.
 *
 * <p>The COBOL paragraphs map as follows:
 * <ul>
 *   <li>{@code 0000-ACCTFILE-OPEN} / {@code 9000-ACCTFILE-CLOSE} - {@link #run}</li>
 *   <li>{@code 1000-ACCTFILE-GET-NEXT} - {@link AccountFileReader}</li>
 *   <li>{@code 1100-DISPLAY-ACCT-RECORD} - {@link #displayAccountRecord}</li>
 *   <li>{@code 9910-DISPLAY-IO-STATUS} / {@code 9999-ABEND-PROGRAM} - {@link #abend}</li>
 * </ul>
 *
 * <p>Amount fields are {@link java.math.BigDecimal}: the {@code S9(10)V99} scale of
 * two is preserved end to end and no value ever passes through a floating point type.
 */
public final class Cbact01c {

    private static final String SEPARATOR = "-------------------------------------------------";

    /** Width of the label column of {@code 1100-DISPLAY-ACCT-RECORD}. */
    private static final int LABEL_WIDTH = 24;

    /** File status set by the COBOL runtime when the dataset is not available. */
    private static final String STATUS_OPEN_FAILED = "35";

    /** File status set by the COBOL runtime for a malformed record. */
    private static final String STATUS_READ_FAILED = "04";

    private Cbact01c() {
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.err.println("usage: Cbact01c <acctfile>");
            System.exit(2);
        }
        try {
            run(Path.of(args[0]), System.out);
        } catch (AbendException e) {
            System.out.flush();
            System.exit(AbendException.ABEND_CODE);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    /**
     * Run the batch job over {@code acctFile}, writing the report to {@code out}.
     *
     * @return the accounts read
     */
    public static List<AccountRecord> run(Path acctFile, PrintStream out) throws IOException {
        line(out, "START OF EXECUTION OF PROGRAM CBACT01C");

        if (!Files.isReadable(acctFile)) {
            line(out, "ERROR OPENING ACCTFILE");
            abend(out, STATUS_OPEN_FAILED, "unable to open " + acctFile);
        }

        List<String> rawRecords = AccountFileReader.readRawRecords(acctFile);
        List<AccountRecord> accounts = new ArrayList<>(rawRecords.size());
        for (String raw : rawRecords) {
            AccountRecord account;
            try {
                account = AccountRecord.parse(raw);
            } catch (IllegalArgumentException e) {
                line(out, "ERROR READING ACCOUNT FILE");
                abend(out, STATUS_READ_FAILED, e.getMessage());
                throw new AssertionError("unreachable");
            }
            accounts.add(account);
            displayAccountRecord(account, out);
            line(out, account.toRawRecord());
        }

        line(out, "END OF EXECUTION OF PROGRAM CBACT01C");
        return accounts;
    }

    /** {@code 1100-DISPLAY-ACCT-RECORD}. */
    static void displayAccountRecord(AccountRecord account, PrintStream out) {
        display(out, "ACCT-ID", account.acctId());
        display(out, "ACCT-ACTIVE-STATUS", account.activeStatus());
        display(out, "ACCT-CURR-BAL", AccountRecord.display(account.currBal()));
        display(out, "ACCT-CREDIT-LIMIT", AccountRecord.display(account.creditLimit()));
        display(out, "ACCT-CASH-CREDIT-LIMIT", AccountRecord.display(account.cashCreditLimit()));
        display(out, "ACCT-OPEN-DATE", account.openDate());
        display(out, "ACCT-EXPIRAION-DATE", account.expirationDate());
        display(out, "ACCT-REISSUE-DATE", account.reissueDate());
        display(out, "ACCT-CURR-CYC-CREDIT", AccountRecord.display(account.currCycCredit()));
        display(out, "ACCT-CURR-CYC-DEBIT", AccountRecord.display(account.currCycDebit()));
        display(out, "ACCT-GROUP-ID", account.groupId());
        line(out, SEPARATOR);
    }

    private static void display(PrintStream out, String label, String value) {
        line(out, label + " ".repeat(LABEL_WIDTH - label.length()) + ":" + value);
    }

    /** Terminate every report line with a newline regardless of the host platform. */
    private static void line(PrintStream out, String text) {
        out.print(text);
        out.print('\n');
    }

    /** {@code 9910-DISPLAY-IO-STATUS} followed by {@code 9999-ABEND-PROGRAM}. */
    private static void abend(PrintStream out, String fileStatus, String detail) {
        line(out, "FILE STATUS IS: NNNN00" + fileStatus);
        line(out, "ABENDING PROGRAM");
        throw new AbendException(detail, fileStatus);
    }
}
